{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 16 - Class Project Part 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genetic variation in long-term experimental warming plots\n",
    "\n",
    "This year's class project will be based on results that my labs received this past year from genome sequencing of small pools of bacterial cells. The first part of the project will consist of classifying the 16S sequences, extract the phylum you are working on and making a high quality phylogenetic tree(s). For more more information see the Nature Communications Giant Virus paper on the Moodle site and the presentation slides.\n",
    "\n",
    "## Determining taxonomic relationships using the RDP classifier\"\n",
    "\n",
    "The use of 16S rRNA massive sequencing has deeply improved the technical possibilities to describe the taxonomic composition and functionality of microbial communities. Following the reduction in DNA sequencing cost, many studies have been performed using amplicon libraries to taxonomically describe microbial communities in many different environments. The large number of sequence reads can be taxonomically assigned by comparison with taxonomically classified sequences present in dedicated databases of 16S rRNA genes, as for instance SILVA, Greengenes or the Ribosomal Database Project. In particular, one of the most popular tools used to assign sequence reads to the prokaryotic taxonomy is the Naïve Bayesian Classifier tool hosted by Ribosomal Database Project (RDP Classifier). The RDP Classifier tool uses a very fast algorithm, based on the Bayes' theorem, suitable for the analysis of large amount of sequence data. \n",
    "\n",
    "The current hierarchy model used by the RDP naïve Bayesian rRNA classifier comes from that proposed in the new phylogenetically consistent higher-order bacterial taxonomy with some minor changes for lineage with few cultivated members. From the highest to the lowest, the major formal taxonomic ranks are: domain, phylum, class, order, family, genus and species. There are occasional intermediate ranks such as “subclass” and “suborder”. Note that confidence values are assigned at each level.\n",
    "\n",
    "## Running the RDP classifier\n",
    "\n",
    "The RDP classifier tool main page can be found at http://rdp.cme.msu.edu/classifier/classifier.jsp  The interactive classifier tool on the RDP web page is best for examining a small number of sequences to obtain taxonomic assignments It only allows the upload of a single sequence file) The advantages of using this version include the fast turn-around time and the rich browsing options over the results. \n",
    "\n",
    "Download the input files for this tutorial from the Moodle site. The sample input file contains over 900 16S rDNA sequences assembled from pools of 50 flow cytometry sorted cells\n",
    "\n",
    "To run the Classifier, select the gene \"16S rRNA\" from the drop-down menu, then you have two choices to submit the test file: Upload one FASTA, GenBank or EMBL formatted sequence file or Cut and paste input sequences. After the job is complete. Click on \"Assignment Detail\" in the upper right corner, then download fixed rank result.  The \"fixrank\" format only outputs the results for a list of selected ranks in the following order: domain, phylum, class, order, family and genus. In case of missing ranks in the lineage, the bootstrap value and the taxon name from the immediate lower rank will be reported. This eliminates the gaps in the lineage, but also introduces non-existing taxon name and rank. \n",
    "\n",
    "### Extracting the 16S rRNA sequences for the taxa for your project.\n",
    "\n",
    "Following assembly of the genomes resulting from flow cytometry sorting of the cells, there were 912 nearly full length 16S rRNA genes. On the Moodle site is a file with your assigned genomic taxa based on the fixrank file. Write a program that extracts your 16S rRNA sequences from the Project_16S_seqs.fna file on the Moodle site. In this project input your taxa on the command line (It should be a generic program that could extract any taxa). Use the fixrank file and the Project_16S_seqs.fna from the minimetagenomics results as input files.\n",
    "\n",
    "### Getting an appropriate group of reference sequences for phylogentic analysis\n",
    "\n",
    "We will compare our sequences to cultured bacterial representatives to provide a phylogenetic context. Go to the RDP Heirarchical brower - http://rdp.cme.msu.edu/hierarchy/hb_intro.jsp. Check the following boxes\n",
    "\n",
    "* Strain:\t Type\n",
    "* Source:\t Isolates\n",
    "* Size:\t >1200\n",
    "* Quality:\t Good\n",
    "* Taxonomy:\t Nomenclatural\n",
    "\n",
    "\n",
    "Find your taxa, then click on the box on the left changing the + to a -. Then click download on the upper right corner. Select \n",
    "\n",
    "* fasta format\n",
    "* remove all gaps\n",
    "\n",
    "Then download the sequence.\n",
    "\n",
    "### Get an outgroup to root your tree\n",
    "\n",
    "You will also need to add an outgroup to your reference tree data set. To do this identify a closely related taxonomic group and get a sequence from a taxa in that group (e.g. if your taxa is at the family level get a sequence from a different family).\n",
    "\n",
    "### Create common fasta headers with just the IDs\n",
    "\n",
    "Write a program that strips all information from the fasta header except the sequence ID (e.g. Ga0207044_1014.65 or S002035495). It is ok if you have two separate programs for the minimetagenomic fasta sequences and the RDP fasta sequences. \n",
    "\n",
    "### Phylogenetic Analysis\n",
    "\n",
    "Merge your reference sequence file with you minimetagenomics results file. Do a phylogenetic analysis as per Lab 10 on CIPRES. \n",
    "\n",
    "* Multiple Sequence Alignment - To preserve the full length (instead of having it truncated to 10 characters) Use Muscle - Under Advanced Parameters - Specify output formats - Fasta output.  Save the output.fasta file to your data folder\n",
    "* Phylogenetic Analysis - For simplicity in interpreting the results - Use FastTreeMP. You will need to click on Parameter Set and Save even if you don't change the parameters. Download the fastree_result.tre to your computer\n",
    "\n",
    "### Pretting up your tree \n",
    "\n",
    "Your trees will only have the IDs in the taxon label. It would be much better to see the taxonomic name for the references sequences. Write a program that substitutes the ID for the taxonomic name (and to avoid the possibility of different sequences have the same taxa name adding the ID) (e.g. Anaerorhabdus_furcosa_S002035495) for the reference sequences.  Make sure to remove spaces or substitute with an underscore or other character. Using as input the newick tree file and the original references sequence fasta file (with the file header).\n",
    "\n",
    "* In Dendroscope - When loading the file select \"Interpret as edge labels (such as bootstrap labels)\"\n",
    "* In Figtree - When loading the file Label nodes as bootstrap. To show the bootstrap values Click on node labels then display bootstrap\n",
    "\n",
    "### Preliminary Report\n",
    "\n",
    "This part of your project report should be in journal article format with \n",
    "\n",
    "* Background (2-3 paragraphs)\n",
    "* Methods (just your analysis methods)\n",
    "* Results (including tree images)\n",
    "* Discussion (putting your findings in the context of known informatin on your taxa)\n",
    "\n",
    "Your should also include as Appendices your programs.\n",
    "\n",
    "1. A program that extracts your 16S rRNA sequences from the Project_16S_seqs.fna file on the Moodle site. In this project input your taxa on the command line (thus it will be a generic program that could extract any taxa). Use the fixrank file and the Project_16S_seqs.fna from the minimetagenomics results as input files.\n",
    "\n",
    "2. A program that strips all information from the fasta header except the sequence ID (e.g. Ga0207044_1014.65 or S002035495). The file name should be input on the command line.\n",
    "\n",
    "3. A program that substitutes the ID for the taxonomic name (e.g. Anaerorhabdus furcosa) for the reference sequences. Using as input the newick tree file and the original references sequence fasta file (with the file header).\n",
    "\n",
    "Submit your project as a Jupyter html file."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
