{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 9: Sequence Alignment and Phylogenetic Analysis\n",
    "\n",
    "\n",
    "## Learning Objectives\n",
    "\n",
    "* Sequence Alignment\n",
    "\n",
    "## Sequence Alignment Background\n",
    "\n",
    "Sequence Alignment is one of the most fundamental operations of bioinformatics. Sequence alignments are used to decide if genes or proteins are related functionally, structurally and evolutionarily.  It is also at the core of sequence search methods such as BLAST.  \n",
    "\n",
    "There are two principal methods of pair-wise sequence alignment\n",
    "\n",
    "* Dot matrix analysis.\n",
    "* The dynamic programming (or DP) algorithm.\n",
    "* Word or k-tuple methods (such as those used in BLAST and FASTA - which we will discuss later)\n",
    "\n",
    "### Dot matrix analysis\n",
    "\n",
    "This method involves constructing a matrix with one of the sequences to be compared running horizontally across the bottom, and the other running vertically along the left-hand side. Each entry of the matrix is a\n",
    "measure of similarity of those two residues on the horizontal and vertical sequence. The simplest scoring system, distinguishes only between identical (dots) and non-identical (blank) residues. \n",
    "\n",
    "The dot plot displays graphically any possible sequence alignments as diagonals on the matrix.\n",
    "Insertions, deletions, direct repeats and inverted repeats can be visually detected on a dot plot.   Dot plot can also detect regions of RNA that are self-complementary and thus might form secondary structure.\n",
    "\n",
    "\n",
    "<img src=\"https://bcrc.bio.umass.edu/courses/spring2012/micbio/micbio660/sites/default/files/dotplot.gif\" width=\"500\" alt=\"Dot plot\" /><br>\n",
    "\n",
    "Ref:<a href=\"http://mummer.sourceforge.net/manual/\"></a>\n",
    "\n",
    "This dot plot represents an alignment of two different strains of Helicobacter pylori. Forward matches are shown in red, while reverse matches are shown in green. This alignment clearly shows a major inversion event centered around the origin of replication.\n",
    "\n",
    "### The dynamic programming algorithm\n",
    "\n",
    "\"Dynamic programming was formalized by mathematician Richard Bellman, who was working at RNAD corporation on optimal decision processes. He wanted to concoct an impressive name that would shield his work from US Secretary of Defense Charles Wilson, a man know to be hostile to mathematics research. His work involved times series and planning - thus 'dynamic' and 'programming' (note, nothing particularly to do with computer programming). Bellman especially liked 'dynamic' because \"it's impossible to use the work dynamic in the pejorative sense\"; he figured dynamic programming was \"something not even a Congressman could object to\" (Bellman 1984).\" (Eddy 2004)\n",
    "\n",
    "The dynamic programming algorithm is at the heart of many bioinformatics programs including BLAST (sequence searching), FASTA (sequence searching), CLUSTALW (multiple sequence alignmnet), HMMER (profile searching), GENSCAN (gene finding), MFOLD (RNA folding), and PHYLIP (phylogenetic analysis). Dynamic programing was first used for global alignment of protein sequences by Needleman and Wunnsch (1970) and to find local alignments by Smith and Waterson (1981).\n",
    "\n",
    "<pre>\n",
    "  MLIGKVIGSVVSTRKN-LNVGKFMIVEPLK\n",
    "  |     |  |  ||   || ||||     |        Global Alignment\n",
    "  MVAARLI-VWANTRKTSLNLGKFMLAEIIK\n",
    "  \n",
    "              TRKN-LNVGKFM\n",
    "              ||   || ||||               Local Alignment\n",
    "              TRKTDLNLGKFML\n",
    "</pre>\n",
    "\n",
    "\n",
    "### Subsitution penalties differ for particular nucleotides or aminoacids\n",
    "\n",
    "Amino acids are put together into proteins based on their three nucleotide codons, and most mutational events usually only change one nucleotide at time.  For example, starting with the Alanine codon GCU and looking at all possible changes on a http://en.wikipedia.org/wiki/Genetic_code#RNA_codon_table \n",
    "\n",
    "    GCU -> UCU = Alanine (A) -> Serine (S)\n",
    "    GCU -> ACU = Alanine (A) -> Threonine (T)\n",
    "    GCU -> CCU = Alanine (A) -> Proline (P)\n",
    "    GCU -> GUU = Alanine (A) -> Valine (V)\n",
    "    GCU -> GAU = Alanine (A) -> Aspartic Acid (D)\n",
    "    GCU -> GGU = Alanine (A) -> Glycine (G)\n",
    "    GCU -> GCA = Alanine (A) -> Alanine (A)\n",
    "    GCU -> GCC = Alanine (A) -> Alanine (A)\n",
    "    GCU -> GCG = Alanine (A) -> Alanine (A)\n",
    "\n",
    "Thus, from the Alanine codon GCU with a single substitution we can only get to 6 of the 19 other amino acids and 33% of the changes are going to result in keeping Alanine at the site in the protein, the other six amino acids would result 11% of the time.  Even if we look at all codons for Alanine, there are still 12 amino acids that can not be made by single substitution.  Thus, the frequency of changing Alanine to another amino acid by a single nucleotide change would be \n",
    "\n",
    "<pre>\n",
    "       A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V\n",
    "    A .33  0   0  .06  0   0  .06 .11  0   0   0   0   0   0  .11 .11 .11  0   0  .11\n",
    "</pre>\n",
    "\n",
    "### Origin of protein substitution matrices and bioinformatics\n",
    " \n",
    "Protein substitution matrices were first developed by Margaret Dayhoff, one of the founders of the field of bioinformatics.   She was particularly interested in the possibility of deducing the evolutionary relationships of organisms from protein sequences. Toward these ends she collected all the known protein sequences and, as a service to the scientific community, made them available to others in 1965 in a small book, the first Atlas of Protein Sequence and Structure.   \n",
    "\n",
    "Ref: http://en.wikipedia.org/wiki/Margaret_Oakley_Dayhoff\n",
    "\n",
    "From these sequences, she and her coworkers developed a model of protein evolution which resulted in the development of a set of widely used substitution matrices. These are frequently called Dayhoff or PAM (Percent Accepted Mutation) matrices and similar matrices are used in many areas of bioinformatics including: sequence similarity searches (e.g. BLAST), multiple sequence alignment tools (e.g ClustalW), phylogenetics and identifying functional regions of proteins.\n",
    "\n",
    "### Natural selection governs which amino acid changes are observed\n",
    "\n",
    "Just like mutation mucks with the frequency matrix so does Natural Selection.  Changes that occur between amino acids that have different biochemical properties are likely to affect the function of the protein.  Therefore, a substitution is more likely to occur between amino acids with similar biochemical properties.  In the above example with the Alanine codon a substitution that yields an amino acid change would result in mostly changes to other neutral amino acids, whereas the frequency of change to Aspartic Acid would probably be much lower. Amazingly, the genetic code has evolved to minimize changes between amino acids with different biochemical properties.\n",
    "\n",
    "<pre>\n",
    "Amino Acid          Side chain polarity Side chain acidity or basicity\n",
    " Alanine              nonpolar            neutral\n",
    " Arginine             polar                  strongly basic\n",
    " Asparagine           polar                  neutral\n",
    " Aspartic acid        polar                  acidic\n",
    " Cysteine             polar                  neutral\n",
    " Glutamic acid        polar                  acidic\n",
    " Glutamine            polar                  neutral\n",
    " Glycine              nonpolar               neutral\n",
    " Histidine            polar                  weakly basic\n",
    " Isoleucine           nonpolar               neutral\n",
    " Leucine              nonpolar               neutral\n",
    " Lysine               polar                  basic\n",
    " Methionine           nonpolar               neutral\n",
    " Phenylalanine        nonpolar               neutral\n",
    " Proline              nonpolar               neutral\n",
    " Serine               polar                  neutral\n",
    " Threonine            polar                  neutral\n",
    " Tryptophan           nonpolar               neutral\n",
    " Tyrosine             polar                  neutral\n",
    " Valine               nonpolar               neutral\n",
    "</pre>\n",
    "\n",
    "### Accounting for multiple substitutions\n",
    "\n",
    "As time goes on and sequence divergence gets larger it gets harder to account for multiple substitutions at the same amino acid position.  PAM protein matrices are based on global alignments of closely related proteins.  However, sequence changes over long evolutionary time scales are not well approximated by compounding small changes that occur over short time scales.  For comparing more distantly related sequence other types of matrices are used.  One of the most common is the The BLOSUM (BLOck SUbstitution Matrix) series of matrices created by Steve Henikoff and colleagues.  Both the PAM and BLOSUM are not expressed as transformation frequencies but the probabilities of transformation are expressed by log-odds scores as shown below for a BLOSUM62 matrix.\n",
    "  \n",
    "\n",
    "<img src=\"https://bcrc.bio.umass.edu/courses/spring2012/micbio/micbio660/sites/default/files/400px-blosum62.gif\" width=\"700\" alt=\"BLOSUM62\">\n",
    "\n",
    "\n",
    "Ref: http://en.wikipedia.org/wiki/Substitution_matrix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multiple Sequence Alignment\n",
    "\n",
    "Why is multiple seqence alignment difficult?\n",
    "\n",
    "If sequences are all the same length, alignment is trivial:\n",
    "\n",
    "        KNMITGAAQMDGAILVVAATDGPMPQTREHVLLARQVEVP\n",
    "        KNMITGAAQMDGAILVVSATDGAMPQTKEHILLARQVGVP\n",
    "        KNMITGAAQMDGAILVVSATDGAMPQTKEHILLARQVGVP\n",
    "        KNMITGAAQMDGAILVVSAADGPMPQTREHILLARQVGVP\n",
    "\n",
    "\n",
    "This sequence alignment is unambiguous because there is no length variation among the sequences. No indels are needed to make the alignment, and the ungapped sequences can simply be arranged together.  However, if the sequenes are of various lengths, problem is more complex, potentially very complex:\n",
    "\n",
    "        RGSALKAVEAPNDPNHEA......YKPIQELLDAMDN.....YIPDPQRDVDKPFL\n",
    "        RGSALKALEGDAAYIEKVR..........ELMQAVDD.....NIPTPEREIDKPFL\n",
    "        RGSALKALE.....IEKVR..........ELMQAGDAAYVDDNIPTPEREIDKPFL\n",
    "        RGSALLALEEMHKNPKTKRGENEWVDKIWELLDAIDE.....YIPTPVRDVDKPFL\n",
    "        RGSALLALEQMHRNPKTRRGENEWVDKIWELLDAIDE.....YIPTPVRDVDKPFL\n",
    "        KGSALQALEALQANPKTARGEDKWVDRIWELLDAVDS.....YIPTPERATDKTFL\n",
    "        RGTARNALESPSKDIN....APEY.KCILELMNAVDE.....YIPTPQRAVDQPFL\n",
    "        KGSALQALE....NAE....DEEKTKCIWELLQAMDD.....YIPAPERDIDKPFL\n",
    "        KGSAFGAMS....NPE....DPESTKCVKELLESMDN.....YFDLPERDIDKPFL\n",
    "        RGSAFAAMS....KPD....DPAATKCLDELLDTMDK.....YFVIPERALDKPFL\n",
    "\n",
    "\n",
    "In many cases the best position to place an indel is ambiguous.  Ideally, one would know the phylogeny for the sequences; this would help infer the sequence of indels.  Unfortunately one normally needs a multiple sequence alignment to make inferences about how the sequences are related.  Most alignment algorithms make a quick approximation of phylogeny, and then base alignments on these.  Sound circular?  You are right and this is a challenging problem that is at the forefront of research in phylogenetics...the joint estimation of the alignment and phylogeny. For this class we will stick to the traditional method of first aligning sequences followed by phylogenetic analysis.\n",
    "\n",
    "Progressive alignment methods are efficient enough to implement on a large scale for many (100s to 1000s) sequences. Progressive alignment services are commonly available on publicly accessible web servers so users need not locally install the applications of interest. The most popular progressive alignment method has been the Clustal family Different portals or implementations can vary in user interface and make different parameters accessible to the user. Clustal Omega is used extensively for phylogenetic tree construction.\n",
    "\n",
    "The basic steps in Clustal are:\n",
    "\n",
    "1. Calculate all possible pairwise alignments, record the score for each pair\n",
    "2. Calculate a guide tree based on the pairwise distances (algorithm: Neighbor Joining)\n",
    "3. Find the two most closely related sequences\n",
    "4. Align these sequences (algorithm: Smith-Waterman).\n",
    "    a. Calculate a consensus of this alignment\n",
    "    b. Replace the two sequences with the consensus\n",
    "    c. Find the two next-most closely related sequences (one of these could be a previously determined consensus sequence).\n",
    "    d. Iterate until all sequences have been aligned\n",
    "5. Expand the consensus sequences with the (gapped) original sequences\n",
    "6. Report the multiple sequence alignment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Software for sequence alignment\n",
    "\n",
    "There are many tools available for sequence alignment.  The common tools are hosting at the European Bioinformatics Institute.\n",
    "\n",
    "* <a href=\"http://www.ebi.ac.uk/Tools/psa/\">Pairwise sequence alignment</a>\n",
    "* <a href=\"http://www.ebi.ac.uk/Tools/msa/\">Multiple sequence alignment</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylogenetic Analysis\n",
    "\n",
    "The CIPRES Science Gateway is a public resource for inference of large phylogenetic trees. It is designed to provide researchers with access to large computational resources of the NSF TeraGrid through a simple browser interface. The CIPRES Science Gateway  The CIPRES Portal V 1.0 permitted users to run popular sequence alignment tools ClustalW and Muscle and the community tree inference tools GARLI, RAxML, PAUP, and MrBayes. \n",
    "\n",
    "Read through the <a href=\"http://www.phylo.org/tools/flash/cipresportal2_data_management.htm\" target=\"_blank\">CIPRES Tutorial</a>.  \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercises \n",
    "\n",
    "1. Align and do phylogenetic analysis on the Core Tree of Life sequences (Lab 13) in CIPRES. Put the resulting tree file in your notebook\n",
    "\n",
    "* Multiple Sequence Alignment - To preserve the full length of the fasta header (instead of having it truncated to 10 characters) Use Muscle - Under Advanced Parameters - Specify output formats - Fasta output.  Save the output.fasta file to your data folder\n",
    "* Phylogenetic Analysis - For simplicity in interpreting the results - Use FastTreeMP. You will need to click on Parameter Set and Save even if you don't change the parameters. Download the fastree_result.tre to your computer\n",
    "\n",
    "2. Download one of the popular treeviewers. <a href=\"http://tree.bio.ed.ac.uk/software/figtree/\" target=\"_blank\">FigTree</a> Take some time to read through the manual.  Upload an tree image produced from Figtree to your Jupyter notebook. In rectangular format\n",
    "\n",
    "* In Figtree - When loading the file Label nodes as bootstrap. To show the bootstrap values Click on node labels then display bootstrap\n",
    "\n",
    "3. Color the tree according to the domains of life. Upload a circular version of the tree to your notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Next - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV5/blob/master/EvoGenV5_Lab15.ipynb\">Lab 15 : Tree Visualization</a>\n",
    "* Previous - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV5/blob/master/EvoGenV5_Lab13.ipynb\">Lab 13 : Introduction to Taxonomic Classification and Phylogenetic Analysis</a> "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
