{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12.1 Introduction to Ontologies\n",
    "\n",
    "From <a href=\"http://en.wikipedia.org/wiki/Ontology\">Wikipedia</a> \"In philosophy, ontology is the most fundamental branch of metaphysics. It studies being or existence and their basic categories and relationships, to determine what entities and what types of entities exist. Ontology thus has strong implications for conceptions of reality.\"  In computer science the term ontology has a more specific meaning.  \"...an ontology is a data model that represents a domain and is used to reason about the objects in that domain and the relations between them.  Ontologies are used in artificial intelligence, the semantic web, and software engineering as a form of knowledge representation about the world or some part of it. Ontologies generally consist of:\n",
    "\n",
    "* Individuals: the basic or \"ground level\" objects.\n",
    "* Classes: sets, collections, or types of objects.\n",
    "* Attributes: properties, features, characteristics, or parameters that objects can have and share.\n",
    "* Relations: ways that objects can be related to one another.\" - <a href=\"http://en.wikipedia.org/wiki/Ontology_%28computer_science%29\"> Wikipedia</a>\n",
    "\n",
    "### Ontologies in Bioinformatics\n",
    "\n",
    "Ontologies in molecular biology are derived from the metabolic pathways database efforts which first attempted to capture the metabolic \"relations\" between genes (and gene products) and to group them into \"classes\" reflecting metabolic pathways. Metabolic pathway databases have proven extremely useful in the annotation and interpretation of genomic data due to our solid understanding of metabolism and the evolutionary conservation of pathways.  As an example, the metabolic pathway analysis system WIT (“What is there?”) at Argonne National Labs was used to help annotate the complete genome sequence of the methanogenic archaeon, ''Methanococcus jannaschii''.  The genes from Methanococcus were mapped to a generic reference metabolic pathway, and genes missing from the pathway were searched for again in the sequence files to determine whether they were incorrectly predicted or misannotated.  If the pathway is still incomplete, it suggests that this particular pathway and metabolic phenotype is not valid for the organism.  Thus, the process of conceptual model building (e.g a metabolic pathway model) and testing with genomic data has proven to be useful for predicting phenotypes of many organisms, although \"What is there?\" is no longer there (it went commercial and then died).  <a href=\"http://biocyc.org/\"> BioCyc</a> and <a href=\"https://www.genome.jp/kegg/\"> KEGG</a> are examples of databases with metabolic roots that have grown into more complete biological information sources in part because of dedicated groups of developers and curators that keep the databases up to date.\n",
    "\n",
    "### The Gene Ontology Project\n",
    "The Gene Ontology project originated by a need to map gene names and functions across the major model eukaryotic organismal databases for <a href=\"http://www.yeastgenome.org/\"> yeast</a>, <a href=\"http://flybase.org/\"> fly</a> and <a href=\"http://www.informatics.jax.org/\">mouse</a>, but has been extended to other organisms.  The Gene Ontology is composed of three structured heirarchical vocabularies:\n",
    "\n",
    "* Cellular Component Ontology:  Subcellular structures, locations, and macromolecular complexes; examples include nucleus, telomere, and origin recognition complex.\n",
    "* Biological Process Ontology:  Broad biological goals, such as mitosis or purine metabolism, that are accomplished by ordered assemblies of molecular functions.\n",
    "* Molecular Function Ontology:  the tasks performed by individual gene products; examples are carbohydrate binding and ATPase activity.\n",
    "\n",
    "Below is a part of the Gene Ontology for the rbcL protein showing the three heirarchical vocabularies\n",
    "\n",
    "<img src=\"RbcL.png\" width=\"500\" alt=\"RbcL\" /> \n",
    "\n",
    "\n",
    "Below is an example of how a gene product (protein) is associated with the Gene Ontology terms in the Arabidopsis database (TAIR).\n",
    "\n",
    "<img src=\"AnnotationDetail.png\" width=\"500\" alt=\"Annotation detail\" /> \n",
    "\n",
    "The GO curators attached genes from a particular species to the appropriate locations in the ontology using several sources of data, including manually sifting through the published literature as well as sequence similarity to a gene with a known function.  This is a treasure trove for the model eukaryotic organisms, but the catch is that if your favorite organism may not have any GO annotation available or just have computationally generated annotation. Read the <a href=\"http://www.geneontology.org/GO.doc.shtml\">Introduction to Gene Ontology</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 12.2 NCBI's Conserved Domain Database (CDD)\n",
    "Identifying conserved protein families and domains are an important part of genome analyses.  NCBI's Conserved Domain Database contains scoring matrices derived from databases of conserved protein domains and proteins families.  These include the curated NCBI Conserved Domains, <a href=\"http://smart.embl-heidelberg.de/\">SMART</a>, <a href=\"http://www.sanger.ac.uk/Software/Pfam/\">PFAM</a>, COGS (Phylogenetic classification of proteins encoded in complete genomes) and KOGS (same as COGS except for eurkaryotes).   The CDD-Search service uses the reverse position-specific BLAST  algorithm (RPSBLAST), that comes with the NCBI blast installation. The query sequence is compared to a position-specific score matrix prepared from the underlying conserved domain alignment.  If you run NCBI's online BLAST a CDD-search is run automatically.  However, you can also down load a local installation of the CDD motif database to generate annotation for your complete genome.  For more information see the <a href=\"http://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml\"> CCD Help Page</a>.\n",
    "\n",
    "## InterPro\n",
    "\n",
    "\"InterPro is a consortium of member databases (PROSITE, Pfam, Prints, ProDom, SMART and TIGRFAMs). Each member database devises methods that can be applied computationally to assign a score for a protein according to how well it matches a given signature. For some types of methods, the classification is binary (i.e. hit or miss), in other cases a numerical value is produced and a cut off point chosen to separate hits from misses. Different member databases create methods/signatures in different ways: some groups build them from alignments studied manually, others use automatic processes with some human input and correction, while ProDom uses an entirely automatic method. See the publications of the different member databases for details.\" <a href=\"http://www.ebi.ac.uk/interpro/\">InterPro home</a>.\n",
    "\n",
    "#### GenBank Record Cphy gbk file\n",
    "\n",
    "<pre>\n",
    "     gene            75..1436\n",
    "                     /locus_tag=\"Cphy_0001\"\n",
    "                     /db_xref=\"GeneID:5742482\"\n",
    "     CDS             75..1436\n",
    "                     /locus_tag=\"Cphy_0001\"\n",
    "                     /note=\"KEGG: cbf:CLI_0001 chromosomal replication\n",
    "                     initiator protein DnaA\n",
    "                     TIGRFAM: chromosomal replication initiator protein DnaA\n",
    "                     PFAM: Chromosomal replication initiator DnaA domain;\n",
    "                     Chromosomal replication initiator DnaA\n",
    "                     SMART: AAA ATPase\"\n",
    "                     /codon_start=1\n",
    "                     /transl_table=11\n",
    "                     /product=\"chromosomal replication initiator protein DnaA\"\n",
    "                     /protein_id=\"YP_001557131.1\"\n",
    "                     /db_xref=\"InterPro:IPR001957\"\n",
    "                     /db_xref=\"InterPro:IPR003593\"\n",
    "                     /db_xref=\"InterPro:IPR013159\"\n",
    "                     /db_xref=\"InterPro:IPR013317\"\n",
    "                     /db_xref=\"GI:160878163\"\n",
    "                     /db_xref=\"GeneID:5742482\"\n",
    "                     /translation=\"MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVI\n",
    "                     KLSVDDTKIGANSLDFIKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAP\n",
    "                     NKIKNESLSYLNPRYTFDTFVVGANNNLAHAASLAVAESPAEIYNPLFIYGGVGLGKT\n",
    "                     HLMHSIAHYILEQNPNSKVLYVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLID\n",
    "                     DIQFIIGKERTQEEFFHTFNTLHESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVD\n",
    "                     IQSPDYETRMAILKKKEELDCLTIDDEVMKYIASNIKSNIRELEGALTKIVALSRLKK\n",
    "                     KEVDVILAEEALKDLISPDNKKTVTLDLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQ\n",
    "                     IAMYLCRKLTSLSLTDIGKMMGNRDHSTVLHGCNKVEKDIKKDPSFQNTIDVLIKKIN\n",
    "                     PTP\"\n",
    "</pre>\n",
    "\n",
    "The gene and locus tag are usually identical for most bacterial genes.  The locus tag refers to the product of the gene.  The gene and locus tag can be different if there are introns or the 5' and/or 3' transcript have been mapped on the gene.\n",
    "\n",
    "The KEGG, TIGRFAM, PFAM, SMART and InterPro fields all results from matches to models (PSSMs, HMMs, etc) in the respective databases.  Here are the links to the InterPro results.\n",
    "\n",
    "* <a href=\"http://www.ebi.ac.uk/interpro/ISearch?query=IPR001957\"> Bac_DnaA </a>\n",
    "* <a href=\"http://www.ebi.ac.uk/interpro/ISearch?query=IPR003593\"> AAA+_ATPase_core </a>\n",
    "* <a href=\"http://www.ebi.ac.uk/interpro/ISearch?query=IPR013159\"> Chromosomal replication initiator, DnaA C-terminal </a>\n",
    "* <a href=\"http://www.ebi.ac.uk/interpro/ISearch?query=IPR013317\"> Chromosomal replication initiator, DnaA </a>\n",
    "\n",
    "#### Arabidopsis gbk file\n",
    "Different genomes have different annotations that depend on the group(s) involved in the sequence and annotation.  In this example we will look a NC_003074.gbk which contains the Arabidopsis chromosome III sequence.  \n",
    "The COMMENT field indicates that this record was curated by TAIR which is the group that is responsible for Arabidopsis community database.  This means that TAIR has added a lot of specific information about the DNA sequence in this record.  This includes the location (to the best of their knowledge) of every gene in this nucleotide.  In the lab last week you found the location of your UNKNOWN DNA sequence in this record and the gene that was present in that region.  Here is an example gene\n",
    "\n",
    "<pre>\n",
    "     gene            complement(21508755..21510653)\n",
    "                     /locus_tag=\"AT3G58060\"\n",
    "                     /db_xref=\"GeneID:824975\"\n",
    "     mRNA            complement(join(21508755..21508931,21509171..21509307,\n",
    "                     21509386..21509536,21509611..21509789,21509903..21510101,\n",
    "                     21510180..21510370,21510452..21510653))\n",
    "                     /locus_tag=\"AT3G58060\"\n",
    "                     /transcript_id=\"NM_115668.2\"\n",
    "                     /db_xref=\"GI:42566021\"\n",
    "                     /db_xref=\"GeneID:824975\"\n",
    "     CDS             complement(join(21508755..21508931,21509171..21509307,\n",
    "                     21509386..21509536,21509611..21509789,21509903..21510101,\n",
    "                     21510180..21510370,21510452..21510653))\n",
    "                     /locus_tag=\"AT3G58060\"\n",
    "                     /note=\"cation efflux family protein / metal tolerance\n",
    "                     protein, putative (MTPc3), member of the cation diffusion\n",
    "                     facilitator (CDF) family, or cation efflux (CE) family,\n",
    "                     PMID:11500563;\n",
    "                     go_component: membrane [goid GO:0016020] [evidence IEA];\n",
    "                     go_component: membrane [goid GO:0016020] [evidence ISS];\n",
    "                     go_function: cation transporter activity [goid GO:0008324]\n",
    "                     [evidence IEA];\n",
    "                     go_function: cation transporter activity [goid GO:0008324]\n",
    "                     [evidence ISS];\n",
    "                     go_function: efflux permease activity [goid GO:0015562]\n",
    "                     [evidence ISS];\n",
    "                     go_process: cation transport [goid GO:0006812] [evidence\n",
    "                     IEA];\n",
    "                     go_process: cation transport [goid GO:0006812] [evidence\n",
    "                     ISS]\"\n",
    "                     /codon_start=1\n",
    "                     /product=\"cation transporter/ efflux permease\"\n",
    "                     /protein_id=\"NP_191365.2\"\n",
    "                     /db_xref=\"GI:42566022\"\n",
    "                     /db_xref=\"GeneID:824975\"\n",
    "                     /translation=\"MEVNYCPETPLLSSNDHEAIDHKPKLTGMVSSMKSNFFADLPQK\n",
    "                     LRSKIDPENPLHLDVSKAAGLKEDEKEYYERQLATLKSFEEVESFLARSDEYTIDEKE\n",
    "                     EEEDRAERAAQELAMQISNWANIFLLALKIYATVKSGSIAIAASTLDSLLDLMAGGIL\n",
    "                     WFTHLSMKNVNIYKYPIGKLRVQPVGIIIFAAVMATLGFQVLLVAAEQLISNEPSEKM\n",
    "                     NHVQLIWLYSIMLSATAIKLVLWIYCKSSRNHIVRAYAKDHHFDVVTNVLGLVAAVLA\n",
    "                     NAFYWWLDPTGAILLAIYTIVNWSGTVMENAVSLIGQSAPPEVLQKLTYLVMRQGGDN\n",
    "                     IKHVDTVRAYTFGVLYFVEVDIELPEDLPLKEAHAIGESLQIKLEELPEVERAFVHLD\n",
    "                     FECHHKPEHSVLSTIPNDL\"\n",
    "</pre>\n",
    "\n",
    "The '''gene''' field tell us the location of the gene in the nucleotide sequence which in this example is on the complement strand from position 21508755 to 21510653.  The locus tag is a unique name for the gene.  The '''CDS''' field (coding sequence) tells us what sequence was used to translate the DNA into a protein.  Since the gene contains a number of introns we would need to join together the sequence coordinates listed in this field to create the splice mRNA that is the template for making proteins.   Although the coding sequence is almost always different from the mRNA, since the mRNA contains untranslated 5' (which contain the ribosomal binding sites) and 3' ends, in GenBank records they are often the same since for many genes the 5' and 3' ends of the mRNA have not been identified.  The '''product''' describes, to the best of our current knowledge or at least to the best of the TAIR curators knowledge at the time this record was made, the function of the protein.  There are several other fields such as go_function and go_process that are derived from the [http://www.geneontology.org Gene Ontology Project].  Mappings from the various databases (PFAM, COG, etc) to GO can be found at \n",
    "http://www.geneontology.org/external2go/. You can write a Python script to extract the GO terms that map to the PFAM and COG terms and add the GO terms to your results!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9.4 Bacterial Microcompartments\n",
    "\n",
    "Today we will run the <i>Clostridium phytofermentans</i> protein sequences against an individual COG and PFAM that represents a bacterial microcompartment protoen to get a feel for RPSBLAST.\n",
    "\n",
    "Bacterial microcompartments (BMCs) are intracellular organelles composed of a proteinacous shell that encapsulates metabolic products that maybe harmful to the cell. These structures are icosahedral in shape, akin to bacterial phages and capsid viruses,\n",
    "but there is no evidence for sequence or structural homology.  Thus, the origin of the BMC shell proteins is an open mystery. BMCs are subject to frequent horizontal gene transfer and loss leading to sporadic distribution across bacteria.\n",
    "\n",
    "The BMCs are frequent encoded by a long operon consisting of ~10-20 genes.  The operons all contains genes for the outer shell proteins of the BMC, One shell protein family is represented by the Protein family (Pfam) database model, (Pfam03319). Proteins having these domains form pentamers which act as the vertices of the polyhedral shell structure and have thus been recently named vertex proteins (purple proteins in the figure below).  The second domain present in shell proteins forms cyclical hexamers that are believed to come together and form the faces of the shell.   This protein domain is represented by the Pfam database model (Pfam00936) and is present in multiple paralogs which number between 3 and 7 genes per BMC locus. We will refer to proteins with this domain as shell face proteins (blue proteins in the figure below), although the location and orientation of all family members has not been experimentally determined.  Using these two protein domains or representative proteins, new BMC shell protein genes are continuously being identified as more bacterial genomes are sequenced.\n",
    "\n",
    "<img src=\"BMC_pathway.png\" width=\"200\" alt=\"BMC pathway\" /> \n",
    "\n",
    "\n",
    "The physiological genes on the BMC operon vary. One group is similar to ethanolamine metabolism genes which contribute to intestinal outbreaks of Salmonella.  A second BMC locus is responsible for the metabolism of fucose, a terminal carbohydrate implicated in bacterial pathogenesis and host immune system response. A putative choline transporter is part of the third BMC locus. The end product of choline metabolism using the BMC is trimethylamine (TMA). Gut microbiota-derived TMA has been implicated in trimethylaminuria and cardiovascular disease and serves as a substrate for gut methanogens. \n",
    "\n",
    "Microcompartments have two features that makes them interesting for comparative  genomics:\n",
    "\n",
    "* Using these two protein domains or representative proteins, new BMC shell protein genes are continuously being identified as more bacterial genomes are sequenced.\n",
    "\n",
    "* All genes are typically organized in a single operon where all functions needed for the microcompartment are present.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Searching the COG and PFAM annotation \n",
    "\n",
    "Today we will run the C. phytofermentans protein sequences (GCF_000018685.1_ASM1868.faa) against an individual PFAM representing the microcompartment vertex protein to get a feel for RPSBLAST.  \n",
    "\n",
    "From Moodle download to your working directory pfam03319.smp\n",
    "\n",
    "NCBI provides a Cog.pn file that works if you are using all COGs in the database, but since we are just using one COG we need to make our own .pn file. Open your text editor and type\n",
    "\n",
    "<pre>\n",
    "    pfam03319.smp \n",
    "</pre>\n",
    "\n",
    "and save the file as pfam03319.pn.  The you will need to format the database as in BLAST.  To do this run\n",
    "\n",
    "<pre>\n",
    "    makeprofiledb -title Pfam.v.26.0 -in pfam03319.pn -out pfam03319 -threshold 9.82 -scale 100.0 -dbtype rps -index true\n",
    "</pre>\n",
    "\n",
    "Then to search all Clostridium phytofermentans proteins (GCF_000018685.1_ASM1868.faa) against this\n",
    "\n",
    "<pre>\n",
    "    rpsblast -query GCF_000018685.1_ASM1868.faa -db pfam03319 -evalue 0.001 -out GCF_000018685.1_ASM1868.faa.pfam03319.results.txt\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the results file many of the C. phytofermentans proteins will have no significant hit, but if you search for \"Sequences producing\" you should find 3 significant hits.  Here is an example of one.\n",
    "\n",
    "<pre>\n",
    "Query= gi|160879333|ref|YP_001558301.1| Ethanolamine utilization\n",
    "protein EutN/carboxysome structural protein Ccml [Clostridium\n",
    "phytofermentans ISDg]\n",
    "         (86 letters)\n",
    "         \n",
    "                                                                   Score    E\n",
    "Sequences producing significant alignments:                      (bits) Value\n",
    "\n",
    "gnl|CDD|66961 pfam03319, EutN_CcmL, Ethanolamine utilisation pro...    70   5e-22\n",
    "\n",
    ">gnl|CDD|66961 pfam03319, EutN_CcmL, Ethanolamine utilisation protein\n",
    "          EutN/carboxysome structural protein Ccml..\n",
    "          Length = 86\n",
    "\n",
    " Score = 70.4 bits (172), Expect = 5e-22\n",
    " Identities = 35/81 (43%), Positives = 49/81 (60%), Gaps = 3/81 (3%)\n",
    "\n",
    "Query: 3  IGKVIGSVVSTRKNENLVGNKFMIV---EPLKSFQENRIVAIDNVGAGIGEYVLVAQGSA 59\n",
    "          I +V G++VSTR+   L G K ++V   +          VA+D VGAG+GE+VL+  GSA\n",
    "Sbjct: 2  IARVRGTLVSTRRIPGLKGKKLLLVVFIDAQGQPLGKYEVAVDPVGAGVGEWVLLVSGSA 61\n",
    "\n",
    "Query: 60 ARIGCGAENSPIDSAIVGIID 80\n",
    "          AR   G E+ P D  ++GIID\n",
    "Sbjct: 62 ARHAAGDESRPTDLTVIGIID 82\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercises\n",
    "\n",
    "Use the complete set of predicted proteins (.faa files) for a representative genomes of <i>Clostridium botulinum A Hall</i>, <i>Hungateiclostridium thermocellum</i>, and <i>Eubacterium rectale</i> that you downloaded for Lab 11 test for the presence of proteins containing pfam03319.\n",
    "\n",
    "1. Which Clostridium botulinum A Hall proteins contain the pfam03319 microcompartment motif?\n",
    "\n",
    "2. Which Hungateiclostridium thermocellum proteins contain the pfam03319 microcompartment motif?\n",
    "\n",
    "3. Which Eubacterium rectale A Hall proteins contain the pfam03319 microcompartment motif?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Next - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV5/blob/master/EvoGenV5_Lab11.ipynb\">Lab 11 : Blast</a>\n",
    "* Previous - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV5/blob/master/EvoGenV5_Lab13.ipynb\">Lab 13 : Phylogenetic analysis</a> "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
