{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 11 : BLAST\n",
    "\n",
    "## Learning Objectives\n",
    "\n",
    "* BLAST fundamentals and terms\n",
    "* Running the NCBI Web BLAST\n",
    "* Set up and run a local version of Blast \n",
    "* Blast one genome against another\n",
    "\n",
    "## 11.1 BLAST\n",
    "\n",
    "The <a href=\"http://www.ncbi.nlm.nih.gov/books/NBK21097/\" target=\"_blank\"> NCBI Tutorial</a> contains an introduction to the basics of BLAST, covering the general approach, the alignment algorithm, how to quantify the comparison of sequences,the insertion of gaps, the statistical aspect of searching and the databases involved.  Key to BLAST's success, both as an algorithm and as a program, has been its use of statistical methods to simultaneously accelerate its search speed and to increase its search sensitivity. In the late 1980's and early 1990's, Altschul and Karlin showed that the significance of pairwise sequence alignments could be rapidly determined or in essence, \"predicted\" using what is called an Extreme Value Distribution. By using this special kind of Poisson distribution and the probabilities that could be calculated from it, Altschul and co-workers demonstrated that it was possible to rapidly identify, assess and extend local sequence blocks or k-tuples more rapidly and more accurately than FASTA or other competing programs.  In class we will go over the tutorial and the <a href=\"http://www.ncbi.nlm.nih.gov/books/NBK62051/\" target=\"_blank\">Glossary of BLAST TERMS</a>\n",
    "\n",
    "\n",
    "### Similarity vs. Identity\n",
    "\n",
    "When protein sequences are aligned, the amino acids at a particular position in the aligment are often said to be identical if they are identical and similar if the amino acids have similar biochemical properties.  For example in the BLAST results below, the V to I comparison is marked by a \"+\".  If you look at the above chart on the biochemical characteristics of proteins both V (valine) and I (Isolecine) are nonpolar neutral amino acids.  Thus, they often can be exchanged in a protein sequence without affecting the protein function.  Protein sequences are often compared based on the % identity which is the fraction of amino acids that are identical in an alignment and the % similarity which is the fraction of similar amino acids.  In the BLAST results the % similarity is the \"Positives\".\n",
    "\n",
    "<pre>\n",
    "  Score =  293 bits (751),  Expect = 7e-78, Method: Composition-based stats.\n",
    "  Identities = 141/202 (69%), Positives = 169/202 (83%), Gaps = 1/202 (0%)\n",
    "    \n",
    "    Query  208  QLISNEPSEKMNHVQLIWLYSIMLSATAIKLVLWIYCKSSRNHIVRAYAKDHHFDVVTNV  267\n",
    "                QL+ N+  EKM   QLIWLYSIMLSAT +KL L+IYC+SS N IV+AYAKDH+FDVVTNV\n",
    "    Sbjct  1    QLVENKAGEKMTPEQLIWLYSIMLSATVVKLALYIYCRSSGNSIVQAYAKDHYFDVVTNV  60\n",
    "    \n",
    "    Query  268  LGLVAAVLANAFYWWLDPTGAILLAIYTIVNWSGTVMENAVSLIGQSAPPEVLQKLTYLV  327\n",
    "                +GLVAAVL + F+WW+DP GA+LLA+YTIVNWSGTV ENAV+L+GQ AP ++LQKLTYL \n",
    "    Sbjct  61   VGLVAAVLGDKFFWWIDPVGAVLLAVYTIVNWSGTVYENAVTLVGQCAPSDMLQKLTYLA  120\n",
    "    \n",
    "    Query  328  MRQGGDNIKHVDTVRAYTFGVLYFVEVDIELPEDLPLKEAHAIGESLQIKLEELPEVERA  387\n",
    "                M+     ++ VDTVRAY+FG LYFVEVDIEL ED+ L EAH+IGESLQ K+E+LPEVERA\n",
    "    Sbjct  121  MKH-DPRVRRVDTVRAYSFGALYFVEVDIELSEDMRLGEAHSIGESLQDKIEKLPEVERA  179\n",
    "    \n",
    "    Query  388  FVHLDFECHHKPEHSVLSTIPN  409\n",
    "                FVH+DFE  HKPEH V S +P+\n",
    "    Sbjct  180  FVHVDFESTHKPEHRVRSRLPS  201\n",
    "</pre>\n",
    "\n",
    "\n",
    "The BLAST output lists similar sequences ordered by their E (expect) value.  The E value of a match represents the chance that the match occurs in a randomly generated database of the same size and composition.  The closer to 0 the E value is, the less likely it occurred by chance.  Thus the lower the E value the better the match.\n",
    "\n",
    "## Setting up and searching your own local DNA and Protein sequence database.\n",
    "\n",
    "The National Center for Biotechnology Information (NCBI) provides tools and sequence files to allow you to create your own sequence database and search it using  BLAST.  A <a href=\"http://www.ncbi.nlm.nih.gov/books/NBK21097/\" target=\"_blank=\"> BLAST Overview and short tutorial</a> is available at NCBI's Education Site.  \n",
    "You should be familiar with the material on this site.  \n",
    "\n",
    "There are several reasons to create a local sequence database as we will do in this class rather than use the web site.\n",
    "\n",
    "1. It is easier to interpret the results. \n",
    "\n",
    "2. You can run multiple searches easily.\n",
    "* search all proteins from a whole genome against GenBank\n",
    "* search all proteins from a whole genome against another.\n",
    "\n",
    "3. You can incorporate \"unpublished\" data that is not in GenBank into your search.\n",
    "* using your own unpublished genomes\n",
    "* draft genomes available at JGI or Sanger institute that are not in GenBank yet.\n",
    "\n",
    "4. You can make a sequence search part of a larger program.\n",
    "\n",
    "5. You can create a specialized database for testing specific hypotheses.\n",
    "\n",
    "\n",
    "## Installing Standalone Blast: OSX\n",
    "\n",
    "If you have installed ebiotools you should have a version of blast installed.  You may still need to set you environmental variables (see below) so that blast can be run from any directory.  If not\n",
    "\n",
    "To download BLAST go to\n",
    "\n",
    "ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST\n",
    "\n",
    "download the latest version (e.g. ncbi-blast-2.7.1+.dmg as of 10/15/2018)\n",
    "\n",
    "Double-click it, and follow the instructions. This will install the blast suite under /usr/local/ncbi. \n",
    "\n",
    "  \n",
    "\n",
    "## Installing Standalone Blast: Windows\n",
    "\n",
    "NCBI has a tutorial that might be helpful, particularly for Windows 7 https://www.ncbi.nlm.nih.gov/books/NBK279690/\n",
    "\n",
    "To download BLAST go to\n",
    "\n",
    "ftp://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/\n",
    "\n",
    "download the latest version (e.g. 2.7.1 as of 10/28/2017)\n",
    "\n",
    "<pre>\n",
    "   ncbi-blast-2.7.1+-win64.exe\n",
    "</pre>\n",
    "\n",
    "I think it is now set up so that you simply need to click and install.  But if not, make a directory on your C drive called blast.  Move the blast*.exe file to your blast directory. Double click to unzip/unstuff the file.   This will create three directories (bin, data, doc)\n",
    "\n",
    "## Installing Standalone Blast: Linux\n",
    "You only need to do this if you get an error when you type blastall in terminal.  Your computer needs to know where the blast programs are so that we can run them from any directory.  This is called setting the path environmental variable.  To do this type in Terminal.\n",
    "\n",
    "<pre>\n",
    "cd\n",
    "echo \"PATH=$PATH/:/INSERT_YOUR_PATH/ncbi-blast-2.7.1+/bin\" >> ~/.bash_profile\n",
    "\n",
    "for EXAMPLE (use your the path on your computer)\n",
    "\n",
    "echo \"PATH=$PATH/:/home/YOURNAME/AdvGen/ncbi-blast-2.7.1+/bin\" >> ~/.bash_profile\n",
    "\n",
    "</pre>\n",
    "\n",
    "The \"source\" command tells the computer to read a file and do different things depending on what the file says.\n",
    "\n",
    "<pre>\n",
    "    source ~/.bash_profile\n",
    "</pre>\n",
    "\t\n",
    "To test that this worked, we'll make the computer tell us our PATH.\n",
    "\n",
    "<pre>\n",
    "    echo $PATH\n",
    "</pre>\n",
    "\t\n",
    "The output of that last command should have a path like /home/jeff/AdvGenblast-2.5.0+/bin somewhere in it.\n",
    "\n",
    "## Running a BLAST search\n",
    "\n",
    "### Testing to see in blast is correctly installed\n",
    "\n",
    "Firt test to see if blast is working correctly on your computer. Close any existing terminal/cmd prompt windows and then reopen terminal/cmd prompt windows. Go to your working directory/folder (e.g. Bioinformatics) Type\n",
    "<pre>\n",
    "    \n",
    "    blastp\n",
    "    \n",
    "</pre>\n",
    "\n",
    "You should get a message \n",
    "\n",
    "BLAST query/options error: Either a BLAST database or subject sequence(s) must be specified\n",
    "Please refer to the BLAST+ user manual.\n",
    "\n",
    "This is a good message. It means blast is working, you know need to give it the files and database.\n",
    "\n",
    "### Getting sequences\n",
    "\n",
    "In your myblast directory create a \"databases\" folder.  \n",
    "\n",
    "For this example (this are the files from Lab 8) go to ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Lachnoclostridium_phytofermentans/representative/GCF_000018685.1_ASM1868v1 and download the following files and put them into your a directory where you will store your blast databases and results.\n",
    "\n",
    "<pre>\n",
    "      GCF_000018685.1_ASM1868.fna\n",
    "      GCF_000018685.1_ASM1868.faa\n",
    "</pre>\n",
    "\n",
    "The \".fna\" extension stands for Fasta Nucleic Acid, which lets you know that this is a nucleic acid sequence in FASTA format. (You will have to wait a moment between these commands while each sequence downloads).  The \".faa\" stands for FASTA Amino Acid, which, of course, is the collection of amino acid sequences of the requested genome in FASTA format.\n",
    "\n",
    "<pre>\n",
    "      Open the GCF_000018685.1_ASM1868.faa file and cut and paste the first protein sequence (with the fasta header) \n",
    "      into a new file and save it in the myblast directory as testseq.faa\n",
    "</pre>\n",
    "\n",
    "### Formatting a BLAST database\n",
    "\n",
    "We need to format sequences that we want to use as our \"database\" in blast search.  To do this in Terminal or Command Prompt go to the AdvGen directory and type\n",
    "\n",
    "For proteins sequences\n",
    "\n",
    "<pre>\n",
    "    makeblastdb -in GCF_000018685.1_ASM1868.faa -out GCF_000018685.1_ASM1868_BLASTDB -dbtype prot -parse_seqids\n",
    "</pre>\n",
    "    \n",
    "Now type\n",
    "\n",
    "<pre>\n",
    "    ls (Unix) or dir (Windows)\n",
    "</pre>\n",
    "    \n",
    "and you will see files like\n",
    "\n",
    "<pre>\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.phr\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.pin\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.pog\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.psd\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.psi\n",
    "    GCF_000018685.1_ASM1868_PROT_BLASTDB.faa.psq\n",
    "</pre>\n",
    "\n",
    "For nucleotide sequences you need to change the -dbtype to nucl.  Here is an example\n",
    "\n",
    "<pre>\n",
    "    makeblastdb -in GCF_000018685.1_ASM1868.fna -out GCF_000018685.1_ASM1868_NUCL_BLASTDB -dbtype nucl -parse_seqids\n",
    "</pre>\n",
    "\n",
    "This files are needed for the BLAST program.   To run a search of the testseq.faa file against the whole genome in Terminal go to your myblast diretory.  \n",
    "\n",
    "### First blast\n",
    " \n",
    "Let's BLAST the the testfiles with the program \"blastall\". You must use options to tell blastall where the database file and query files you want to use are. Since we are in the same directory as our query file, we can just say the name of the file, other wise we would have to use both the path to the file and the name of the file. So, to blast our amino acid test sequence against the amino acid database, we do:\n",
    "\n",
    "<pre>\n",
    "    \n",
    "    blastp -query testseq.faa -db GCF_000018685.1_ASM1868_PROT_BLASTDB\n",
    "    \n",
    "</pre>\n",
    "\n",
    "The results will be print to the screen in a tab-delimited format. To save it to a file \n",
    "\n",
    "<pre>\n",
    "\n",
    "    blastp -query testseq.faa -db GCF_000018685.1_ASM1868_PROT_BLASTDB -out testseq.blast.txt \n",
    "\n",
    "</pre>\n",
    "\n",
    "The BLAST output lists similar sequences ordered by their E (expect) value.  The E value of a match represents the chance that the match occurs in a randomly generated database of the same size and composition.  The closer to 0 the E value is, the less likely it occurred by chance.  Thus the lower the E value the better the match.  In this search the best matching hit should have the same coordinates that you found using your perl program.\n",
    "\n",
    "Here are a few helpful hints in interpreting BLAST results\n",
    "\n",
    "* Your results depend on the database your are searching.  In this search we expect to find an exact match to the whole length of our sequence.\n",
    "* Since BLAST is a local alignment tool only part of your query sequence may be similar to another sequence in the database.  \n",
    "* It is very possible that your sequence will be similar to other sequences in the database.  This may be because of repetitive sequences (e.g. transposable elements) or that your sequence is evolutionarily related to another sequence in the genome.\n",
    "\n",
    "### BLAST options\n",
    "\n",
    "Since we wanted to compare a protein sequence against a protein database, we used \"blastp\". Here is the full list of available programs:\n",
    "\n",
    "* blastp - compares an amino acid query sequence against a protein sequence database.\n",
    "* blastn - compares a nucleotide query sequence against a nucleotide sequence database.\n",
    "* blastx - compares a nucleotide query sequence translated in all reading frames against a protein sequence database.\n",
    "* tblastn - compares a protein query sequence against a nucleotide sequence database dynamically translated in all reading frames.\n",
    "* tblastx - compares the six-frame translations of a nucleotide query sequence against the six-frame translations of a nucleotide sequence database.\n",
    "\n",
    "The \"-help\" option will list the options available for each program and a brief explanation.\n",
    "\n",
    "The \"-db\" option is the path to the database you want to use. \n",
    "\n",
    "The \"-evalue\" option is used if you want to filter the results that are returned by E value. Using \"-evalue .001\" report rows with E values less than 10e-3.\n",
    "\n",
    "### Output formats \n",
    "The default output from the BLAST website is in html format (the coding language that makes most webpages). However this is difficult to parse and you may want other things like just the sequences returned or a tabular format of the e-values and identity scores that can be copied into your favourite spreadsheet program.\n",
    "The flag for this is -outfmt followed by a number which denotes the format requested. The numbers correspond like so:\n",
    "\n",
    "<pre>\n",
    "\n",
    "0 = pairwise,\n",
    "1 = query-anchored showing identities,\n",
    "2 = query-anchored no identities,\n",
    "3 = flat query-anchored, show identities,\n",
    "4 = flat query-anchored, no identities,\n",
    "5 = XML Blast output,\n",
    "6 = tabular,\n",
    "7 = tabular with comment lines,\n",
    "8 = Text ASN.1,\n",
    "9 = Binary ASN.1\n",
    "10 = Comma-separated values\n",
    "11 = BLAST archive format (ASN.1)\n",
    "\n",
    "</pre>\n",
    "\n",
    "The most common output formats are 0 and 6. 0 is the default and will output will give a table of each query-hit pair with a bit score and e-value followed by the pairwise alignment of each pair.  Custom output formats can be created.  See the manual https://www.ncbi.nlm.nih.gov/books/NBK279675/. for examples  A portion of example output using 0 is:- \n",
    "\n",
    "<pre>\n",
    "Query= gi|160878163|ref|YP_001557131.1| chromosomal replication initiator\n",
    "protein DnaA [Lachnoclostridium phytofermentans ISDg]\n",
    "\n",
    "Length=453\n",
    "                                                                      Score     E\n",
    "Sequences producing significant alignments:                          (Bits)  Value\n",
    "\n",
    "ref|YP_001557131.1|  chromosomal replication initiator protein Dn...    924   0.0  \n",
    "\n",
    "\n",
    ">ref|YP_001557131.1| chromosomal replication initiator protein DnaA [Lachnoclostridium \n",
    "phytofermentans ISDg]\n",
    "Length=453\n",
    "\n",
    " Score =   924 bits (2389),  Expect = 0.0, Method: Compositional matrix adjust.\n",
    " Identities = 453/453 (100%), Positives = 453/453 (100%), Gaps = 0/453 (0%)\n",
    "\n",
    "Query  1    MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSVDDTKIGANSLDF  60\n",
    "            MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSVDDTKIGANSLDF\n",
    "Sbjct  1    MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSVDDTKIGANSLDF  60\n",
    "\n",
    "Query  61   IKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAPNKIKNESLSYLNPRYTFD  120\n",
    "            IKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAPNKIKNESLSYLNPRYTFD\n",
    "Sbjct  61   IKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAPNKIKNESLSYLNPRYTFD  120\n",
    "\n",
    "Query  121  TFVVGANNNLAHAASLAVAESPAEIYNPLFIYGGVGLGKTHLMHSIAHYILEQNPNSKVL  180\n",
    "            TFVVGANNNLAHAASLAVAESPAEIYNPLFIYGGVGLGKTHLMHSIAHYILEQNPNSKVL\n",
    "Sbjct  121  TFVVGANNNLAHAASLAVAESPAEIYNPLFIYGGVGLGKTHLMHSIAHYILEQNPNSKVL  180\n",
    "\n",
    "Query  181  YVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLIDDIQFIIGKERTQEEFFHTFNTL  240\n",
    "            YVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLIDDIQFIIGKERTQEEFFHTFNTL\n",
    "Sbjct  181  YVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLIDDIQFIIGKERTQEEFFHTFNTL  240\n",
    "\n",
    "Query  241  HESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEELDCLTI  300\n",
    "            HESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEELDCLTI\n",
    "Sbjct  241  HESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEELDCLTI  300\n",
    "\n",
    "Query  301  DDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDNKKTVTL  360\n",
    "            DDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDNKKTVTL\n",
    "Sbjct  301  DDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDNKKTVTL  360\n",
    "\n",
    "Query  361  DLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMGNRDHST  420\n",
    "            DLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMGNRDHST\n",
    "Sbjct  361  DLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMGNRDHST  420\n",
    "\n",
    "Query  421  VLHGCNKVEKDIKKDPSFQNTIDVLIKKINPTP  453\n",
    "            VLHGCNKVEKDIKKDPSFQNTIDVLIKKINPTP\n",
    "Sbjct  421  VLHGCNKVEKDIKKDPSFQNTIDVLIKKINPTP  453\n",
    "</pre>\n",
    "\n",
    "Here is option 6\n",
    "\n",
    "<pre>\n",
    "   blastp -query testseq.faa -db NC_010001_PROT_BLASTDB -evalue 1e-10 -outfmt 6\n",
    "   gi|160878163|ref|YP_001557131.1|\tgi|160878163|ref|YP_001557131.1|\t100.00\t453\t0\t0\t1\t453\t1\t453\t0.0\t  924\n",
    "</pre>\n",
    "\n",
    "Where the columns are \n",
    "<pre>\n",
    "Query id\n",
    "Subject id\n",
    "% identity\n",
    "alignment length\n",
    "mismatches\n",
    "gap\n",
    "openings\n",
    "q. start\n",
    "q. end\n",
    "s. start\n",
    "s. end\n",
    "e-value\n",
    "bit score\n",
    "</pre>\n",
    "\n",
    "### Running multiple BLAST queries\n",
    "\n",
    "To run multiple blast queries simply put the fasta queries in one file and each one will be BLASTed individually.  As an example open the NC_010001.faa file and cut and paste the first 5 protein sequences (with the fasta header) into a new file and save it in the myblast directory as testseq5.faa. Now try\n",
    "\n",
    "<pre>\n",
    "    blastp -query testseq5.faa -db NC_010001_PROT_BLASTDB -out testseq5_results.txt\n",
    "</pre>\n",
    "\n",
    "All 5 blast results will appear in one file.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercises\n",
    "\n",
    "1. Using the <a href =\"http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastHome\" target=\"_blank\">NCBI online blast program</a>, BLASTP the cellulase gene from <i>Clostridium thermocellum</i> - http://www.ncbi.nlm.nih.gov/protein/289859?report=fasta - against the <i>C. phytofermentans gene</i> genome by specifying under the organism Clostridium phytofermentans (taxid:66219).  What is the evidence that C. phytofermentans does or does not contains a cellulase gene?  Are there multiple copies of this gene? If so which one(s)?  Does the <i>C. phytofermentans gene</i> include similar annotated protein domains?   Include your blast results file.\n",
    "\n",
    "2. BLASTP the human p53 gene - http://www.ncbi.nlm.nih.gov/protein/23491729?report=fasta against the database of non-redundant protein sequences using the NCBI web site. Under algorithm parameters select 1000 max target sequenes.  Once the results are ready click on Taxonomy reports at the top of the page.  Are there homologs of the human gene present in fishes?  \n",
    "\n",
    "3. In <i>Clostridium botulinum</i> a human neurotoxins causes the flaccid muscular paralysis seen in botulism. Using the local version of Blast, find the neurotoxin A gene in a database of 3 other genomes.  1) download the fasta file for the toxin protein BLASTP this toxin gene - https://www.ncbi.nlm.nih.gov/protein/P0DPI1.1?report=fasta.  2) Download the complete set of predicted proteins (.faa files) for a representative genomes of <i>Clostridium botulinum A Hall</i>, <i>Hungateiclostridium thermocellum</i>, and <i>Eubacterium rectale</i> from ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/. Get the .faa file from the latest versions and the top file in that directory (if there are more than one).  3) Create a database of multiple sequences\n",
    "\n",
    "    In Linux and Mac OSX\n",
    "    <pre>\n",
    "       cat file1.faa file2.faa file3.faa > all3.faa\n",
    "    </pre>\n",
    "\n",
    "    For those using the DOS Prompt, the following format works\n",
    "\n",
    "    <pre>\n",
    "       copy file1.faa+file2.faa+file3.faa all3.faa\n",
    "    </pre>\n",
    "\n",
    "    Remember to format the database before running Blast. Use the default output format to save your results for Lab 12\n",
    "    \n",
    "\n",
    "### Setting your Environmental Variables on Windows\n",
    "On some computers there are issues with the local path to blast. The local of the path (e.g \"C:\\blast\\data\") must be the same as the place where you install the blast.exe above.  This can be done be going in the windows \"Start\" \"control panel\" then open systems (some versions of xp this is under control panel > Performance and Maintence > System) or directly to System.  Then click on the \"advanced\" tab then on \"environmental variables\" then edit Path and add a semicolon and the path to your blast bin folder.  Put this text in your Path \n",
    "\n",
    "<pre>\n",
    "\n",
    "C:\\blast\\bin.\n",
    "\n",
    "</pre>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Next - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV5/blob/master/EvoGenV5_Lab12.ipynb\">Lab 12 : RPS Blast</a>\n",
    "* Previous - <a href=\"http://nbviewer.ipython.org/github/jeffreyblanchard/EvoGenV4/blob/master/EvoGenV5_Lab5.ipynb\">Lab 10 : Input, Output and Working with Fasta Files</a> "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
